From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Sun, 22 Mar 2026 03:12:38 +0000
Subject: [PATCH] Fix item meta inconsistencies

The order of the set of data components hidden via the
HIDE_ADDITIONAL_TOOLTIP is not consistent in paper as it uses a plain
set over items with non defined hashcodes.

As the mojang implementation of the TooltipDisplay makes use of a
LinkedSet for storing the data components, this leads to non consistent
set order for tooltip displays across different JVMs.

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index b623a882a1bd04380e379a2662ab77e9247ceeff..001472fe39da2ae301cd9bc1a6c736ca2f4ca1cf 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -316,7 +316,7 @@ public class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDat
     private List<net.minecraft.advancements.predicates.BlockPredicate> canBreakPredicates;
 
     // hide_additional_tooltip backward compatibility based on TooltipDisplayComponentFix#CONVERTED_ADDITIONAL_TOOLTIP_TYPES
-    private static final Set<DataComponentType<?>> HIDDEN_COMPONENTS_PREVIOUSLY = Set.of(
+    private static final Set<DataComponentType<?>> HIDDEN_COMPONENTS_PREVIOUSLY = new ReferenceLinkedOpenHashSet<>(List.of( // KTP - consistent order of hide additional tooltip entries
         DataComponents.BANNER_PATTERNS,
         DataComponents.BEES,
         DataComponents.BLOCK_ENTITY_DATA,
@@ -335,7 +335,7 @@ public class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDat
         DataComponents.POTION_CONTENTS,
         DataComponents.TROPICAL_FISH_PATTERN,
         DataComponents.WRITTEN_BOOK_CONTENT
-    );
+    )); // KTP - consistent order of hide additional tooltip entries
 
     private static final Map<ItemFlag, Set<DataComponentType<?>>> ITEM_FLAG_EQUIVALENTS = ImmutableMap.<ItemFlag, Set<DataComponentType<?>>>builder()
         .put(ItemFlag.HIDE_ATTRIBUTES, Set.of(DataComponents.ATTRIBUTE_MODIFIERS))
